media/js/base/site.js

```diff
@@ -162,5 +193,64 @@
 
         // Add class to reflect javascript availability for CSS
         h.className = h.className.replace(/\bno-js\b/, 'js');
+    }
+
+    function getHighEntropyFromUAString() {
+        window.site.platformVersion = window.site.getPlatformVersion();
+        window.site.archType = window.site.getArchType();
+        window.site.archSize = window.site.getArchSize();
+    }
+
+    (function () {
+        window.site.platform = window.site.getPlatform();
+
+        // For browsers that support client hints, get high entropy
+        // values that might be frozen in the regular user agent string.
+        if (
+            'userAgentData' in navigator &&
+            typeof navigator.userAgentData.getHighEntropyValues === 'function'
+        ) {
+            navigator.userAgentData
+                .getHighEntropyValues([
+                    'architecture',
+                    'bitness',
+                    'platformVersion'
+                ])
+                .then(function (ua) {
+                    if (ua.platformVersion) {
+                        if (window.site.platform === 'windows') {
+                            window.site.platformVersion =
+                                window.site.getWindowsVersionCH(
+                                    ua.platformVersion
+                                );
+                        } else {
+                            window.site.platformVersion = ua.platformVersion;
+                        }
+                    }
+
+                    if (ua.architecture) {
+                        window.site.archType = ua.architecture;
+                    }
+
+                    if (ua.bitness) {
+                        window.site.archSize = window.site.getArchSizeCH(
+                            ua.bitness
+                        );
+                    }
+
+                    updateHTML();
```

Comment:
Does `updateHTML()` raise an exception in failure? I'm assuming it does, but if not could we have a single call to it after the if/else then/catch blocks? it looks like it's always called, just with different data sources

Reply:
@stevejalim JS promises are async, so moving to a single call after all the code would probably result in this being run too early.